utils/spdx/src/test/kotlin/SpdxExpressionTest.kt

```diff
@@ -446,6 +446,26 @@ class SpdxExpressionTest : WordSpec({
         "not contain duplicate valid choice different left and right expressions" {
             "a AND a AND b".toSpdx().validChoices() should containExactly("a AND b".toSpdx())
         }
+
+        "work for nested AND expressions" {
+            val license = "(MIT OR GPL-2.0-only) AND (MIT OR BSD-3-Clause OR GPL-1.0-or-later) AND " +
+                "(MIT OR BSD-3-Clause OR GPL-2.0-only)"
+
+            val choices = license.toSpdx().validChoices().map { it.toString() }
+
+            choices should containExactlyInAnyOrder(
+                "MIT AND MIT AND MIT",
+                "MIT AND MIT AND BSD-3-Clause",
+                "MIT AND MIT AND GPL-2.0-only",
+                "MIT AND BSD-3-Clause AND GPL-2.0-only",
+                "MIT AND GPL-1.0-or-later AND MIT",
+                "MIT AND GPL-1.0-or-later AND BSD-3-Clause",
+                "MIT AND GPL-1.0-or-later AND GPL-2.0-only",
+                "GPL-2.0-only AND BSD-3-Clause AND BSD-3-Clause",
+                "GPL-2.0-only AND GPL-1.0-or-later AND BSD-3-Clause",
+                "GPL-2.0-only AND GPL-1.0-or-later AND GPL-2.0-only"
```

Comment:
Should this not contain all possible permutations? Some are missing like "MIT AND BSD-3-Clause AND MIT" or "MIT AND BSD-3-Clause AND GPL-2.0-only".

Reply:
I was at first wondering the same, but `validChoice()` returns a `Set`, and it is `"MIT AND BSD-3-Clause AND MIT".toSpdx() == "MIT AND MIT AND BSD-3-Clause".toSpdx()`, so these permutations are not added to the set.

Reply:
Which indeed is unfortunate, as it's not clear to the user which permutation will be contained in the set. However, that's required as `isValidChoice()` uses `choice in validChoices()` which does *not* check for semantic equality.

Reply:
That is, I would actually propose to change

https://github.com/oss-review-toolkit/ort/blob/488027d8deb92829212178bc720bdc7b6de443ee/utils/spdx/src/main/kotlin/SpdxExpression.kt#L181

to

    fun isValidChoice(choice: SpdxExpression): Boolean = !choice.offersChoice() && validChoices().any { it == choice }

Reply:
Ok, so if this change fixes the problem that the choices are wrong this is already a big improvement, but it seems we need to do more work for a nice UX. For example, "MIT AND MIT AND MIT" should also get collapsed to just "MIT".

Reply:
> For example, "MIT AND MIT AND MIT" should also get collapsed to just "MIT".

You can already use "MIT" instead of "MIT AND MIT AND MIT", see the [test I've added in the last commit](https://github.com/oss-review-toolkit/ort/pull/8106/commits/3892bd3aefb84f8533a6d3a403166b11222d46a3#diff-4e0a232e1e0a7388198459bfe18eed7f326560f25c9ea8782b7da41912bacd9bR495) of this PR.

Reply:
> You can already use "MIT" instead of "MIT AND MIT AND MIT", see the [test I've added in the last commit](https://github.com/oss-review-toolkit/ort/pull/8106/commits/3892bd3aefb84f8533a6d3a403166b11222d46a3#diff-4e0a232e1e0a7388198459bfe18eed7f326560f25c9ea8782b7da41912bacd9bR495) of this PR.

I was also thinking of being able to show the valid choices to the user, but that is about a future improvement anyway.

Reply:
> Could you add one more test for an expression with deeper nesting, like (A OR (B AND (C OR D))) AND (E OR F)? 

@sschuberth Are you planning to add such a test? I think it would be useful but if you don't want to I don't insist.

Reply:
I'm looking into it and meanwhile discovered another thing I'd like to double-check.